src/client/java/net/neoforged/neoforge/client/event/RegisterParticleGroupsEvent.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.client.event;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import net.minecraft.client.particle.Particle;
+import net.minecraft.client.particle.ParticleEngine;
+import net.minecraft.client.particle.ParticleGroup;
+import net.minecraft.client.particle.ParticleRenderType;
+import net.neoforged.bus.api.Event;
+import net.neoforged.fml.LogicalSide;
+import net.neoforged.fml.event.IModBusEvent;
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * Fired for registering additional {@linkplain ParticleGroup particle group factories}.
+ * <p>This is used when creating particle groups for particles in {@link ParticleEngine#createParticleGroup(ParticleRenderType)}.
+ *
+ * <p>This event is fired on the mod-specific event bus,
+ * only on the {@linkplain LogicalSide#CLIENT logical client}.</p>
+ */
+public class RegisterParticleGroupsEvent extends Event implements IModBusEvent {
+    private final Map<ParticleRenderType, Function<ParticleEngine, ParticleGroup<?>>> particleGroupFactories;
+    private final List<ParticleRenderType> particleRenderOrder;
+
+    @ApiStatus.Internal
+    public RegisterParticleGroupsEvent(
+            Map<ParticleRenderType, Function<ParticleEngine, ParticleGroup<?>>> particleGroupFactories,
+            List<ParticleRenderType> particleRenderOrder) {
+        this.particleGroupFactories = particleGroupFactories;
+        this.particleRenderOrder = particleRenderOrder;
+    }
+
+    /**
+     * Registers a factory function for the given {@linkplain Particle#getGroup() particle group}, to be rendered
+     * before all currently registered types.
+     *
+     * @param particleRenderType An identifier uniquely identifying the particle group.
+     * @param factory            A factory function used to create a {@link ParticleGroup} for the particle group.
+     *
+     * @throws IllegalArgumentException when {@code particleRenderType} has already been registered.
+     */
+    public void registerBeforeAll(ParticleRenderType particleRenderType, Function<ParticleEngine, ParticleGroup<?>> factory) {
```

Comment:
I don't think there is much value in providing custom ordering of the groups. Everything should be using `RenderPipeline`s, either directly or indirectly though `RenderType`s, now and therefore set up and tear down its state fully without relying on state from previous types like vanilla did in the past. The order only needs to be consistent, so relying on registration order is sufficient.

Reply:
It does influence draw order, yes, but I don't think you can solve the ordering issue with a primitive solution like this, you'd need a broader grouping of render types to sort them against each other without them knowing about each other. In my eyes that's already a lost cause though due to vanilla "quad particles" being used for anything from a tiny falling leaf to a large explosion shock wave, the latter of which would almost fit your example of a volumetric effect.